fendermint/vm/interpreter/src/bytes.rs

```diff
@@ -31,6 +31,73 @@ impl<I> BytesMessageInterpreter<I> {
     }
 }
 
+#[async_trait]
+impl<I> ProposalInterpreter for BytesMessageInterpreter<I>
+where
+    I: ProposalInterpreter<Message = ChainMessage>,
+{
+    type State = I::State;
+    type Message = Vec<u8>;
+
+    /// Parse messages in the mempool and pass them into the inner `ChainMessage` interpreter.
+    async fn prepare(
+        &self,
+        state: Self::State,
+        msgs: Vec<Self::Message>,
+    ) -> anyhow::Result<Vec<Self::Message>> {
+        // We could include a flag to disable this parsing if we know that the inner interpreter
+        // does not care about proposals, but in our case we know that we'll eventually want to
+        // consult the IPLD Resolver.
+
+        let mut chain_msgs = Vec::new();
+
+        for msg in msgs {
+            match fvm_ipld_encoding::from_slice::<ChainMessage>(&msg) {
+                Err(e) => {
+                    // This should not happen because the `CheckInterpreter` implementation below would
+                    // have rejected any such user transaction.
+                    tracing::warn!(
+                        error = e.to_string(),
+                        "failed to decode message in mempool as ChainMessage"
+                    );
+                }
+                Ok(msg) => chain_msgs.push(msg),
+            }
+        }
+
+        let chain_msgs = self.inner.prepare(state, chain_msgs).await?;
```

Comment:
Not sure for bottom up, but for top down, it's not needed to deserialize the bytes to get `ChainMessage` as no inspection needed. Or it's likely someone will send a `IPCMessage` instead of a `SignedMessage` and we have to inspect here?

Reply:
Added a mode in https://github.com/consensus-shipyard/fendermint/pull/185/commits/22e3719c92bb05bc5598abfb2294155909160324